### PR TITLE
Update sample code with stable alloc crate

### PR DESCRIPTION
**Pull request by lzutao**
Based on #170 .

### CODE REVIEW

File: src/vec-alloc.md [hunk 1]
```diff
@@ -157,46 +166,45 @@ such we will guard against this case explicitly.
...
+                old_num_bytes <= (std::isize::MAX as usize) / 2,
+                "capacity overflow"
+            );
 
             let new_num_bytes = old_num_bytes * 2;
-            let ptr = heap::reallocate(self.ptr.as_ptr() as *mut _,
-                                        old_num_bytes,
-                                        new_num_bytes,
-                                        align);
+            let ptr = alloc::realloc(self.ptr.as_ptr() as *mut u8, layout, new_num_bytes);
```
Comment (Gankra) [inline src/vec-alloc.md]: This isn't correct. The layout you need to pass here is for an array, but this is the layout for a single value. You must use from_size_align_unchecked or from_size_align (generally the former is more appropriate since we're doing our own error-handling).

File: src/vec-insert-remove.md [hunk 2]
```diff
@@ -16,17 +16,20 @@ using the old len.
 pub fn insert(&mut self, index: usize, elem: T) {
     // Note: `<=` because it's valid to insert after everything
     // which would be equivalent to push.
-    assert!(index <= self.len, "index out of bounds");
-    if self.cap == self.len { self.grow(); }
+    assert!(self.len >= index, "index is out of bounds");
```
Comment (Gankra) [inline src/vec-insert-remove.md]: Why did you invert this assertion? This is not standard style.

File: src/vec-insert-remove.md [hunk 3]
```diff
@@ -16,17 +16,20 @@ using the old len.
...
         if index < self.len {
-            // ptr::copy(src, dest, len): "copy from source to dest len elems"
-            ptr::copy(self.ptr.offset(index as isize),
-                      self.ptr.offset(index as isize + 1),
-                      self.len - index);
+            ptr::copy(
+                self.ptr.as_ptr().offset(index as isize),
+                self.ptr.as_ptr().offset(index as isize + 1),
+                self.len - index,
+            );
```
Comment (Gankra) [inline src/vec-insert-remove.md]: Why did you remove this comment?

File: src/vec-insert-remove.md [hunk 4]
```diff
@@ -38,13 +41,18 @@ Remove behaves in the opposite manner. We need to shift all the elements from
 ```rust,ignore
 pub fn remove(&mut self, index: usize) -> T {
     // Note: `<` because it's *not* valid to remove after everything
-    assert!(index < self.len, "index out of bounds");
+    assert!(self.len > index, "index is out of bounds");
+
```
Comment (Gankra) [inline src/vec-insert-remove.md]: Another changed assertion.

File: src/vec-insert-remove.md [hunk 5]
```diff
@@ -38,13 +41,18 @@ Remove behaves in the opposite manner. We need to shift all the elements from
...
-                  self.ptr.offset(index as isize),
-                  self.len - index);
+        let result = ptr::read(self.ptr.as_ptr().offset(index as isize));
+
+        ptr::copy(
+            self.ptr.as_ptr().offset(index as isize + 1),
+            self.ptr.as_ptr().offset(index as isize),
+            self.len - index,
+        );
+
```
Comment (Gankra) [inline src/vec-insert-remove.md]: Why are you reformatting the code?

File: src/vec-into-iter.md [hunk 6]
```diff
@@ -134,12 +133,8 @@ impl<T> Drop for IntoIter<T> {
         if self.cap != 0 {
             // drop any remaining elements
             for _ in &mut *self {}
-
-            let align = mem::align_of::<T>();
-            let elem_size = mem::size_of::<T>();
-            let num_bytes = elem_size * self.cap;
             unsafe {
-                heap::deallocate(self.buf.as_ptr() as *mut _, num_bytes, align);
+                alloc::dealloc(self.buf.as_ptr() as *mut _, alloc::Layout::new::<T>());
```
Comment (Gankra) [inline src/vec-into-iter.md]: Another place with an incorrect Layout.

File: src/vec-raw.md [hunk 7]
```diff
@@ -18,45 +18,47 @@ struct RawVec<T> {
...
-
 impl<T> Drop for RawVec<T> {
     fn drop(&mut self) {
         if self.cap != 0 {
-            let align = mem::align_of::<T>();
-            let elem_size = mem::size_of::<T>();
-            let num_bytes = elem_size * self.cap;
             unsafe {
-                heap::deallocate(self.ptr.as_mut() as *mut _, num_bytes, align);
+                alloc::dealloc(self.ptr.as_ptr() as *mut u8, alloc::Layout::new::<T>());
```
Comment (Gankra) [inline src/vec-raw.md]: Another incorrect Layout

File: src/vec-raw.md [hunk 8]
```diff
@@ -116,15 +118,20 @@ impl<T> Drop for IntoIter<T> {
 impl<T> Vec<T> {
     pub fn into_iter(self) -> IntoIter<T> {
         unsafe {
-            // need to use ptr::read to unsafely move the buf out since it's
+            // need to use `ptr::read` to unsafely move the buf out since it's
```
Comment (Gankra) [inline src/vec-raw.md]: These backticks aren't really helpful inside comments, since the markdown is never rendered.

File: src/vec-raw.md [hunk 9]
```diff
@@ -116,15 +118,20 @@ impl<T> Drop for IntoIter<T> {
...
             IntoIter {
-                start: *buf.ptr,
-                end: buf.ptr.offset(len as isize),
+                start: buf.ptr.as_ptr(),
+                end: if buf.cap == 0 {
+                    // can't offset off this pointer, it's not allocated!
+                    buf.ptr.as_ptr()
+                } else {
+                    buf.ptr.as_ptr().offset(len as isize)
+                },
```
Comment (Gankra) [inline src/vec-raw.md]: I'm pretty sure this is unnecessary, but it's fair this isn't documented properly.

https://github.com/rust-lang/rust/issues/65108

File: src/vec-zsts.md [hunk 10]
```diff
@@ -35,55 +35,57 @@ method of RawVec.
...
 
     fn grow(&mut self) {
         unsafe {
-            let elem_size = mem::size_of::<T>();
-
-            // since we set the capacity to usize::MAX when elem_size is
+            // Since we set the capacity to `usize::MAX` for ZST is
             // 0, getting to here necessarily means the Vec is overfull.
-            assert!(elem_size != 0, "capacity overflow");
+            assert!(mem::size_of::<T>() != 0, "capacity overflow");
```
Comment (Gankra) [inline src/vec-zsts.md]: You should make this check the layout.size(), but I guess that doesn't matter because you probably need to redo this code because this is the wrong layout in the general case.

File: src/vec-zsts.md [hunk 11]
```diff
@@ -35,55 +35,57 @@ method of RawVec.
...
             } else {
-                let new_cap = 2 * self.cap;
-                let ptr = heap::reallocate(self.ptr.as_ptr() as *mut _,
-                                            self.cap * elem_size,
-                                            new_cap * elem_size,
-                                            align);
+                let new_cap = self.cap * 2;
+                let ptr = alloc::realloc(
+                    self.ptr.as_ptr() as *mut u8,
+                    layout,
```
Comment (Gankra) [inline src/vec-zsts.md]: Another wrong layout.

File: src/vec-zsts.md [hunk 12]
```diff
@@ -35,55 +35,57 @@ method of RawVec.
...
-        // don't free zero-sized allocations, as they were never allocated.
-        if self.cap != 0 && elem_size != 0 {
-            let align = mem::align_of::<T>();
-
-            let num_bytes = elem_size * self.cap;
+        // Do not free zero-sized allocations, as they were never allocated.
+        if self.cap != 0 && mem::size_of::<T>() != 0 {
             unsafe {
-                heap::deallocate(self.ptr.as_ptr() as *mut _, num_bytes, align);
+                alloc::dealloc(self.ptr.as_ptr() as *mut u8, alloc::Layout::new::<T>());
```
Comment (Gankra) [inline src/vec-zsts.md]: Another incorrect layout.

**Comment by CreepySkeleton** [comment]
@lzutao could you please also replace `Unique` with `NonNull` so it's finally works on stable?

**Comment by lzutao** [comment]
@CreepySkeleton Thanks for the interest. 
I don't think NonNull could replace Unique. Could you elaborate more about that?
If you want to read about Unique, here is the link: https://doc.rust-lang.org/nightly/src/core/ptr/unique.rs.html#8

**Comment by CreepySkeleton** [comment]
@lzutao 
What's similar between `NonNull` and `Unique`
- They both cannot be `null`
- Both are pointers
- Both are covariant over `T`

`Unique`'s features `NonNull` lacks:
- `notNull` is neither `Send` nor `Sync` because it can be aliased. But the implementation guarantees that the pointer won't ever be aliased so I believe we can just do this
  ```rust
  unsafe impl<T: Send> Send for Vec<T> {}
  unsafe impl<T: Sync> Sync for Vec<T> {}
```

*AFAIK*, that's all. As we can see, in this case `Unique` *can* be replaced by `NonNull` very well.

**Comment by clifton** [comment]
ill close #182 as its a duplicate (although just covering all aspects of the vec example) but may be worth looking to see if theres anything worth adding from there.

it would be good to merge in something like this, as i was not able to get any of the examples to compile on nightly and im sure others have the same problem.

**Comment by lzutao** [comment]
I have no interest in this PR.
